### PR TITLE
enhanced IIS manage task

### PR DESCRIPTION
**Pull request by rajatagrawal-dev**

### CODE REVIEW

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/README_IISAppMgmt.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# WinRM - IIS Web App Management
```
Comment (asranja) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/README_IISAppMgmt.md]: Readme has to be updated.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/AppCmdOnTargetMachines.ps1 [hunk 2]
```diff
@@ -0,0 +1,925 @@
...
+        
+        [string]$AppCmdCommands
+    )
+
+    $physicalPathAuthCredentials = Get-PSCredentials -username $PhysicalPathAuthUserName -password $PhysicalPathAuthUserPassword
+    $appPoolCredentials = Get-PSCredentials -username $AppPoolUsername -password $AppPoolPassword
+    
+    $bindingsArray = ""
+    if (![string]::IsNullOrWhiteSpace($Bindings)) {
+        $bindingsArray = $Bindings | ConvertFrom-Json
```
Comment (asranja) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/AppCmdOnTargetMachines.ps1]: ConvertFrom-Json is available from PS versions 2 and above. Is there any risk ?

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV1/Main.ps1 [hunk 3]
```diff
@@ -30,9 +30,10 @@ param (
     [string]$deployInParallel
```
Comment (asranja) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV1/Main.ps1]: Change the data type(string to int) of iisVersion parameter in Add-SslCert function of AppCmdOnTargetMachines.ps1 file of V1 task.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/ManageIISWebApp.ps1 [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+                        sslThumbprint = Test-SSLCertificateThumbprint -sslCertThumbPrint $sslCertThumbPrint -ipAddress $ipAddress -protocol $protocol -port $port ;
+                        hostname = Get-Hostname -port $port -hostNameWithSNI $hostNameWithSNI -hostNameWithHttp $hostNameWithHttp -hostNameWithOutSNI $hostNameWithOutSNI -sni $serverNameIndication ;
+                    })
+                }
+                else {
+                    $bindingsArray = Validate-Bindings -bindings $bindings
+                }
+            }
+
+            $bindingsJson = $bindingsArray | ConvertTo-Json
```
Comment (asranja) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/ManageIISWebApp.ps1]: Same comment as above for ConvertFrom-Json.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/task.json [hunk 5]
```diff
@@ -0,0 +1,749 @@
+{
+  "id": "0F5CD14F-3C01-4D5C-8F7A-EB96C5738DCC",
+  "name": "IISWebAppMgmt",
+  "friendlyName": "WinRM - IIS Web App Management",
+  "description": "Connect via WinRM, to create or update a Website, Web App, Virtual Directories, and Application Pool locally on IIS",
+  "helpMarkDown": "[More Information](http://aka.ms/IISMgmt)",
+  "category": "Deploy",
```
Comment (asranja) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/task.json]: preview: true ?

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/task.json [hunk 6]
```diff
@@ -0,0 +1,749 @@
...
+      "required": true,
+      "helpMarkDown": "Provide the administrator's login of the target machines."
+    },
+    {
+      "name": "AdminPassword",
+      "type": "string",
+      "label": "Password",
+      "defaultValue": "",
+      "required": true,
+      "helpMarkDown": "Provide the administrator's password of the target machines. <br>The best practice is to create a variable in the Build or Release definition, and mark it as 'Secret' to secure it, and then use it here, like '$(adminPassword)'."
```
Comment (asranja) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/task.json]: rename definition to pipeline.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/task.json [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+    },
+    {
+        "name": "WebsiteAuthUserPassword",
+        "type": "string",
+        "label": "Password",
+        "required": false,
+        "groupName": "Website",
+        "defaultValue": "",
+        "visibleRule": "WebsitePhysicalPathAuth = WebsiteWindowsAuth",
+        "helpMarkDown": "Provide the user's password that will be used to access the website's physical path. <br/>The best practice is to create a variable in the Build or Release definition, and mark it as 'Secret' to secure it, and then use it here, like '$(userCredentials)'. <br> Note: Special characters in password are interpreted as per <a href=\"https://go.microsoft.com/fwlink/?linkid=843470\">command-line arguments</a>"
```
Comment (asranja) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/task.json]: same comment as above.

**Review by asranja — APPROVED** [review]

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/AppCmdOnTargetMachines.ps1 (hunk 2 above)
Comment (rajatagrawal-dev) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/AppCmdOnTargetMachines.ps1]: Don't think so. We should be ok here, especially as this a new major version.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/ManageIISWebApp.ps1 (hunk 4 above)
Comment (rajatagrawal-dev) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/ManageIISWebApp.ps1]: Same reply as above for ConvertFrom-Json.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/README_IISAppMgmt.md (hunk 1 above)
Comment (rajatagrawal-dev) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV2/README_IISAppMgmt.md]: Same as rest

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV1/Main.ps1 (hunk 3 above)
Comment (rajatagrawal-dev) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppMgmt/IISWebAppMgmtV1/Main.ps1]: Done

**Pull request merged**